File: backend/src/models/GeofenceEvent.ts
import { Schema, model, Types, Document } from "mongoose";

export interface IGeofenceEvent extends Document {
  _id: Types.ObjectId;
  parentId: Types.ObjectId;
  childId: Types.ObjectId;
  geofenceId: Types.ObjectId;
  geofenceName: string;
  eventType: "EXIT" | "ENTRY";
  zoneType: "SAFE_ZONE" | "RESTRICTED_ZONE";
  latitude: number;
  longitude: number;
  accuracy?: number;
  speed?: number;
  distanceFromCenter: number;
  geofenceRadius: number;
  address?: string;
  isRead: boolean;
  triggeredAt: Date;
  createdAt: Date;
}

const geofenceEventSchema = new Schema<IGeofenceEvent>(
  {
    parentId: { type: Schema.Types.ObjectId, ref: "User", required: true, index: true },
    childId: { type: Schema.Types.ObjectId, ref: "User", required: true, index: true },
    geofenceId: { type: Schema.Types.ObjectId, ref: "Geofence", required: true, index: true },
    geofenceName: { type: String, required: true },
    eventType: { type: String, enum: ["EXIT", "ENTRY"], required: true },
    zoneType: { type: String, enum: ["SAFE_ZONE", "RESTRICTED_ZONE"], required: true },
    latitude: { type: Number, required: true },
    longitude: { type: Number, required: true },
    accuracy: { type: Number, default: null },
    speed: { type: Number, default: null },
    distanceFromCenter: { type: Number, required: true },
    geofenceRadius: { type: Number, required: true },
    address: { type: String, default: "" },
    isRead: { type: Boolean, default: false, index: true },
    triggeredAt: { type: Date, required: true, default: Date.now, index: true },
  },
  { timestamps: true }
);

geofenceEventSchema.index({ childId: 1, triggeredAt: -1 });
geofenceEventSchema.index({ parentId: 1, isRead: 1 });

export const GeofenceEvent = model<IGeofenceEvent>("GeofenceEvent", geofenceEventSchema);

File: parent-app/lib/models/geofence_event.dart
import 'package:latlong2/latlong.dart';

class GeofenceEvent {
  final String id;
  final String parentId;
  final String childId;
  final String geofenceId;
  final String geofenceName;
  final String eventType; // "EXIT" | "ENTRY"
  final String zoneType; // "SAFE_ZONE" | "RESTRICTED_ZONE"
  final double latitude;
  final double longitude;
  final double? accuracy;
  final double? speed;
  final double distanceFromCenter;
  final double geofenceRadius;
  final String address;
  final bool isRead;
  final DateTime triggeredAt;

  GeofenceEvent({
    required this.id,
    required this.parentId,
    required this.childId,
    required this.geofenceId,
    required this.geofenceName,
    required this.eventType,
    required this.zoneType,
    required this.latitude,
    required this.longitude,
    this.accuracy,
    this.speed,
    required this.distanceFromCenter,
    required this.geofenceRadius,
    this.address = '',
    this.isRead = false,
    required this.triggeredAt,
  });

  LatLng get latLng => LatLng(latitude, longitude);

  bool get isExit => eventType == 'EXIT';
  bool get isEntry => eventType == 'ENTRY';
  bool get isRestricted => zoneType == 'RESTRICTED_ZONE';

  factory GeofenceEvent.fromJson(Map<String, dynamic> json) {
    return GeofenceEvent(
      id: (json['_id'] ?? json['id']).toString(),
      parentId: (json['parentId'] ?? '').toString(),
      childId: (json['childId'] ?? '').toString(),
      geofenceId: (json['geofenceId'] ?? '').toString(),
      geofenceName: json['geofenceName'] as String? ?? 'Safety Boundary',
      eventType: json['eventType'] as String? ?? 'EXIT',
      zoneType: json['zoneType'] as String? ?? 'SAFE_ZONE',
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      accuracy: (json['accuracy'] as num?)?.toDouble(),
      speed: (json['speed'] as num?)?.toDouble(),
      distanceFromCenter: (json['distanceFromCenter'] as num?)?.toDouble() ?? 0.0,
      geofenceRadius: (json['geofenceRadius'] as num?)?.toDouble() ?? 200.0,
      address: json['address'] as String? ?? '',
      isRead: json['isRead'] as bool? ?? false,
      triggeredAt: json['triggeredAt'] != null
          ? DateTime.parse(json['triggeredAt'].toString()).toLocal()
          : DateTime.now(),
    );
  }
}
